README.rst
============
PyMODM_Motor
============


An asynchronous ODM on top of `PyMODM`_ ODM library using `Motor`_ asynchronous
Python MongoDB driver. ``PyMODM_Motor`` works on ``Python 3.5`` and up. Some features
such as asynchronous comprehensions require at least ``Python 3.6``. ``PyMODM_Motor``
can be used with `asyncio`_ as well as with `Tornado`_.

``PyMODM_Motor`` uses ``PyMODM``'s machinery whenever it posible and just modify some 
functions and methods that actually work with database. So the most of 
`official PyMODM documentation`_ is fully compatible (see 
`Differences between PyMODM_Motor and PyMODM`_ for details) with ``PyMODM_Motor`` 
and can be browsed to learn more. You can also take a look at the simple 
`blog example`_ built with `asyncio`_ and `AIOHTTP`_ asynchronous web framework.

.. _PyMODM: https://pypi.python.org/pypi/pymodm
.. _Motor: https://pypi.python.org/pypi/motor
.. _official PyMODM documentation: http://pymodm.readthedocs.io/en/stable
.. _blog example: https://github.com/ilex/pymodm_motor/tree/develop/example/blog/aiohttp
.. _asyncio: https://docs.python.org/3/library/asyncio.html
.. _Tornado: https://pypi.python.org/pypi/tornado
.. _AIOHTTP: https://pypi.python.org/pypi/aiohttp


Install
=======

At this moment you need to install ``pymodm`` as follow::

    pip install https://github.com/ilex/pymodm/archive/fix_dereference.zip

And then install ``PyMODM_Motor`` using ``pip``::
    
    pip install https://github.com/ilex/pymodm_motor/archive/develop.zip 


Differences between PyMODM_Motor and PyMODM
===========================================

Although most of the ``PyMODM_Motor`` is very similar to ``PyMODM`` 
there are some differences due to asynchronous nature of ``PyMODM_Motor``:

1) Use classes with ``Motor`` prefix for models, managers and querysets 
   (note that each ``Motor*`` class derived from appropriate ``PyMODM`` class 
   and inherits most of their behavior):
    
   - ``pymodm_motor.MotorMongoModel`` in place of ``pymodm.MongoModel``
   - ``pymodm_motor.MotorEmbeddedMongoModel`` in place of ``pymodm.EmbeddedMongoModel``
   - ``pymodm_motor.MotorMongoModelMetaclass`` in place of ``pymodm.TopLevelMongoModelMetaclass``
   - ``pymodm_motor.MotorManager`` in place of ``pymodm.Manager``
   - ``pymodm_motor.MotorQuerySet`` in place of ``pymodm.QuerySet``
    
2) Some methods and functions are coroutines (see `List of Coroutines`_) so 
   the should be called with ``await``.

   .. code-block:: python

        from pymodm_motor import fields, MotorMongoModel

        class User(MotorMongoModel):
            name = fields.CharField()

        async def handler():
            user = await User(name='Bob').save()
            await user.refresh_from_db()
            await user.delete()
            users_count = await User.objects.count()
            
3) There is no *auto dereferencing* in ``PyMODM_Motor`` use ``dereference`` 
   coroutine explicitly to dereference reference fields in model instance.

   .. code-block:: python
       
        from bson import ObjectId
        from pymodm_motor import fields, MotorMongoModel
        from pymodm_motor.dereference import dereference

        class User(MotorMongoModel):
            # note that library will add an _id field
            # as there is no field with primary_key=True
            name = fields.CharField()

        class Post(MotorMongoModel):
            # note that title is an _id field
            title = fields.CharField(primary_key=True)
            user = fields.ReferenceField(User)

        async def handler():
            user = await User(name='Bob').save()
            post = await Post(title='My post', user=user).save()

            post_from_db = await Post.objects.get({'_id': 'My post'})
            assert isinstance(post_from_db.user, ObjectId) 

            await dereference(post_from_db)
            assert isinstance(post_from_db.user, User)

4) As there is no *auto dereferencing* the ``no_auto_dereference`` context
   manager is useless.
5) There is no *auto indexes creation* in ``PyMODM_Motor`` use 
   ``MotorQuerySet.create_indexes()`` coroutine method explicitly. As a bonus
   you can use ``switch_connection`` and ``switch_collection`` context managers
   with this method.

   .. code-block:: python
        
        from pymodm_motor import fields, MotorMongoModel
        from pymodm_motor.context_managers import swith_collection
        from pymongo import IndexModel, ASCENDING

        class User(MotorMongoModel):
            name = fields.CharField()

            class Meta:
                indexes = [
                    IndexModel([('name', ASCENDING)], unique=True, name='name_index')
                ]

        async def create_indexes():
            for model in (User, ): # list here models classes with indexes
                await model.objects.create_indexes()

        async def app_init():
            await create_indexes()
            # create indexes in another collection
            with switch_collection(User, 'backup_user') as BackupUser:
                await BackupUser.objects.create_indexes()

6) There are some additional parameters in ``connect`` function:

   - ``mongo_driver``: a string constant that specify which of the drivers to use
     ``pymodm_motor.MOTOR_ASYNCIO_DRIVER`` or ``pymodm_motor.MOTOR_TORNADO_DRIVER``.
   - ``kwargs``: will be passed to appropriate MotorClient. For example ``io_loop``
     parameter can be specified to pass a specific loop.

7) As there is no *auto indexes creation* ``connect`` can be called in any place but
   before any db operations are called.
8) For retrieve bunch of objects ``MotorQuerySet`` returns an asynchronous iterator
   or asynchronous generator (for Python 3.6 and up) so to iterate over items use
   ``async for`` construction. As a consequence you can not use ``list(await Model.objects.all())``
   or list comprehension ``[model for model in await Model.objects.all()]``. 
   If you need a list use ``MotorQuerySet.to_list()`` coroutine which returns 
   a list of models instances. For Python 3.6 the recommended way is to use 
   `asynchronous comprehensions`_.

   .. code-block:: python

        from pymodm_motor import fields, MotorMongoModel
        from pymodm_motor.dereference import dereference

        class User(MotorMongoModel):
            name = fields.CharField()

        async def handler():
            # iterate over objects 
            async for user in User.objects.all():
                print(user.name)

            # get list of the objects
            users = await User.objects.all().to_list()

            # get a list of the objects and dereference them
            users = await User.objects.all().to_list(dereference=True)

            # WITH PYTHON 3.6 AND UP

            # Note this will work only with Python 3.6 and up
            # create a list of dereferenced objects 
            users = [await dereference(user) 
                     async for user in User.objects.all()]

            # Note this will work only with Python 3.6 and up
            # create a list of tuples
            users_id_name = [(user._id, user.name) 
                             async for user in User.objects.all()]
                             
            # Note this will work only with Python 3.6 and up
            # create a dict
            users = {user._id: user.name async for user in User.objects.all()}

9) A slice operator and getitem operator should be used with ``await``.
    
   .. code-block:: python

        from pymodm_motor import fields, MotorMongoModel

        class User(MotorMongoModel):
            name = fields.CharField()

        async def handler():
            name = (await User.objects[3]).name
            users = await User.objects[2:3]  # users is a list of Users


.. _asynchronous comprehensions: https://www.python.org/dev/peps/pep-0530/#asynchronous-comprehensions

List of coroutines
==================
These functions and methods are coroutines or return awaitable:

- ``pymodm_motor.dereference`` module:

  - ``pymodm_motor.dereference.dereference``
  - ``pymodm_motor.dereference.dereference_id``

- ``pymodm_motor.MotorMongoModel`` class:

  - ``pymodm_motor.MotorMongoModel.save``
  - ``pymodm_motor.MotorMongoModel.delete``
  - ``pymodm_motor.MotorMongoModel.refresh_from_db``

- ``pymodm_motor.MotorQuerySet`` class:

  - ``pymodm_motor.MotorQuerySet.count``
  - ``pymodm_motor.MotorQuerySet.get``
  - ``pymodm_motor.MotorQuerySet.first``
  - ``pymodm_motor.MotorQuerySet.bulk_create``
  - ``pymodm_motor.MotorQuerySet.delete``
  - ``pymodm_motor.MotorQuerySet.update``
  - ``pymodm_motor.MotorQuerySet.to_list``
  - ``pymodm_motor.MotorQuerySet.create_indexes``
  - ``pymodm_motor.MotorQuerySet.__getitem__``


Example
=======

Here's a basic example of how to define some models and connect them to MongoDB:

.. code-block:: python

    import asyncio
    from pymongo import IndexModel, TEXT, ASCENDING
    from pymodm_motor import (
        connect, fields, MOTOR_ASYNCIO_DRIVER, 
        MotorMongoModel, MotorEmbeddedMongoModel)


    # Now let's define some Models.
    class User(MotorMongoModel):
        # Use 'email' as the '_id' field in MongoDB.
        email = fields.EmailField(primary_key=True)
        fname = fields.CharField()
        lname = fields.CharField()

        class Meta:
            indexes = [IndexModel([('fname', ASCENDING)])]


    class BlogPost(MotorMongoModel):
        # This field references the User model above.
        # it just stores an user's _id in MongoDB
        author = fields.ReferenceField(User)
        title = fields.CharField(max_length=100)
        content = fields.CharField()
        tags = fields.ListField(fields.StringField(max_length=20))
        # These Comment objects will be stored inside each Post document in the
        # database.
        comments = fields.EmbeddedDocumentListField('Comment')

        class Meta:
            # Text index on content can be used for text search.
            indexes = [IndexModel([('content', TEXT)])]

    # This is an "embedded" model and will be stored as a sub-document.
    class Comment(MotorEmbeddedMongoModel):
        author = fields.ReferenceField(User)
        body = fields.CharField()
        vote_score = fields.IntegerField(min_value=0)


    async def create_indexes():
        # create all indexes
        for model in (User, BlogPost):
            await model.objects.create_indexes()


    async def go(loop):
        
        # Connect to MongoDB first. PyMODM_Motor supports all URI options supported by
        # Motor. Make sure also to specify a database in the connection string and 
        # one of the drivers MOTOR_ASYNCIO_DRIVER or MOTOR_TORNADO_DRIVER.
        # You can also specify other parameters to pass them to MotorClient.
        # For example you can specify a loop.
        connect('mongodb://localhost:27017/myApp', 
                mongo_driver=MOTOR_ASYNCIO_DRIVER, io_loop=loop)
        
        # Explicitly create indexes as PyMODM_Motor does not do that automaticaly
        await create_indexes()

        # We need to save these objects before referencing them later.
        han_solo = await User(
            'mongoblogger@reallycoolmongostuff.com', 'Han', 'Solo').save()
        chewbacca = await User(
            'someoneelse@reallycoolmongostuff.com', 'Chewbacca', 'Thomas').save()

        post = await BlogPost(
            # Since this is a ReferenceField, we had to save han_solo first.
            author=han_solo,
            title="Five Crazy Health Foods Jabba Eats.",
            content="...",
            tags=['alien health', 'slideshow', 'jabba', 'huts'],
            comments=[
                Comment(author=chewbacca, body='Rrrrrrrrrrrrrrrr!', vote_score=42)
            ]
        ).save()

        # Find objects using familiar MongoDB-style syntax.
        slideshows = BlogPost.objects.raw({'tags': 'slideshow'})

        # Only retrieve the 'title' field.
        slideshow_titles = slideshows.only('title')

        # 'Five Crazy Health Foods Jabba Eats.'
        print((await slideshow_titles.first()).title)

    # create an asyncio loop
    loop = asyncio.get_event_loop()
    # run our coroutine
    loop.run_until_complete(go(loop))


License
=======

The library is licensed under Apache License, Version 2.0.


CHANGELOG.rst
Changelog
=========

0.3.1 (XXXX-XX-XX)
------------------

The first ``pymodm_motor`` release.


example/blog/aiohttp/blog.py
"""Blog application using aiohttp and pymodm with motor_asyncio driver."""
import asyncio
import datetime
import functools
import os.path

import aiohttp_jinja2
import aiohttp_session
import jinja2

import blog_models

from aiohttp import web

from bson import ObjectId
from pymodm_motor.errors import ValidationError
from pymongo.errors import DuplicateKeyError

from pymodm_motor import connect, MOTOR_ASYNCIO_DRIVER
from pymodm_motor.dereference import dereference

from blog_models import User, Post, Comment


async def init_app(loop):
    """Initialize aiohttp web application."""
    APP_DIR = os.path.dirname(os.path.abspath(__file__))

    # Establish a connection to the database.
    connect('mongodb://localhost:27017/blog',
            mongo_driver=MOTOR_ASYNCIO_DRIVER, io_loop=loop)
    # pymodm_motor does not create indexes automatically as pymodm does.
    # Use MotorModel.objects.create_indexes() coroutine to create
    # models' indexes explicitly. Application initialization is
    # a good place to do that.
    await blog_models.create_indexes()
    # Create application
    app = web.Application(debug=True, loop=loop)

    # setup jinja2 templates
    aiohttp_jinja2.setup(
        app,
        loader=jinja2.FileSystemLoader(os.path.join(APP_DIR, 'templates')))
    aiohttp_jinja2.get_env(app).filters['human_date'] = human_date

    # setup session
    # SimpleCookieStorage is only for demo
    # it's very non-secure, don't use this storage for real app
    aiohttp_session.setup(
        app,
        aiohttp_session.SimpleCookieStorage())

    # add routes
    app.router.add_get('/', index, name='index')
    app.router.add_get('/login', login, name='login')
    app.router.add_post('/login', login)
    app.router.add_get('/logout', logout, name='logout')
    app.router.add_get('/users/new', new_user, name='new_user')
    app.router.add_post('/users/new', new_user)
    app.router.add_get('/posts/new', create_post, name='create_post')
    app.router.add_post('/posts/new', create_post)
    app.router.add_get('/posts/{post_id}', get_post, name='get_post')
    app.router.add_post('/comments/new', new_comment, name='new_comment')

    # add static route
    # this is only for demo, use web servers to handle static
    app.router.add_static(prefix='/static/',
                          path=os.path.join(APP_DIR, 'static'))

    return app


def human_date(value, format="%B %d at %I:%M %p"):
    """Format a datetime object to be human-readable in a template."""
    return value.strftime(format)


def logged_in(func):
    """Decorator that redirects to login page if a user is not logged in."""
    @functools.wraps(func)
    async def wrapper(*args):
        request = args[-1]
        session = await aiohttp_session.get_session(request)
        if 'user' not in session:
            raise web.HTTPFound(request.app.router['login'].url_for())
        return await func(*args)

    return wrapper


@logged_in
@aiohttp_jinja2.template('new_post.html')
async def create_post(request):
    if request.method == 'GET':
        # return context for template
        return {}
    else:
        session = await aiohttp_session.get_session(request)
        form = await request.post()
        if form['date']:
            post_date = form['date']
        else:
            post_date = datetime.datetime.now()
        try:
            await Post(
                title=form['title'],
                date=post_date,
                body=form['content'],
                author=session['user']).save()
        except ValidationError as exc:
            return {
                'errors': exc.message
            }

        # raise redirect response to avoid template decorator
        raise web.HTTPFound(request.app.router['index'].url_for())


@aiohttp_jinja2.template('login.html')
async def login(request):
    if request.method == 'GET':
        # Return context for login form.
        return {}
    else:
        # Login.
        session = await aiohttp_session.get_session(request)
        form = await request.post()
        email, password = form['email'], form['password']
        try:
            # Note: logging users in like this is acceptable for demonstration
            # projects only.
            user = await User.objects.get(
                {'_id': email, 'password': password})
        except User.DoesNotExist:
            return {
                'error': 'Bad email or password'
            }

        # Store user in the session.
        session['user'] = user.email

        # raise redirect response to avoid template decorator
        raise web.HTTPFound(request.app.router['index'].url_for())


async def logout(request):
    session = await aiohttp_session.get_session(request)
    session.pop('user', None)
    # store flash message between requests
    session['flash'] = 'You have been successfully logged out.'
    return web.HTTPFound(request.app.router['index'].url_for())


@aiohttp_jinja2.template('new_user.html')
async def new_user(request):
    if request.method == 'GET':
        return {}
    else:
        form = await request.post()
        try:
            # Note: real applications should handle user registration more
            # securely than this.
            # Use `force_insert` so that we get a DuplicateKeyError if
            # another user already exists with the same email address.
            # Without this option, we will update (replace) the user with
            # the same id (email).
            await User(
                email=form['email'],
                handle=form['handle'],
                password=form['password']).save(force_insert=True)
        except ValidationError as ve:
            return {
                'errors': ve.message
            }
        except DuplicateKeyError:
            # Email address must be unique.
            return {
                'errors': {
                    'email': [
                        'There is already a user with that email address.'
                    ]
                }
            }

    # raise redirect response to avoid template decorator
    raise web.HTTPFound(request.app.router['index'].url_for())


@aiohttp_jinja2.template('index.html')
async def index(request):
    session = await aiohttp_session.get_session(request)
    # restore flash message if any
    flash = session.pop('flash', None)
    flashed_messages = [flash] if flash else []

    # Use a list here so we can do "if posts" efficiently in the template.
    # Note that pymodm_motor queryset cannot return iterator
    # but asynchronous iterator (to use with async for) or
    # asynchronous generator (for python version >= 3.6)
    # >> posts = Post.objects.all().to_list()
    # is equivalent to:
    # >> posts = list()
    # >> async for item in Post.objects.all():
    # >>     posts.append(item)
    # And with python version >= 3.6 you can use asynchronous comprehensions
    # >> posts = [item async for item in Posts.objects.all()]
    return {
        'posts': await Post.objects.all().to_list(dereference=True),
        'session': session,
        'flashed_messages': flashed_messages
    }


async def get_post(request):
    post_id = request.match_info['post_id']
    try:
        # `post_id` is a string, but it's stored as an ObjectId in the db.
        # Note that auto_dereference is off for pymodm_motoro models and
        # dereference should be used explicitly
        post = await dereference(
            await Post.objects.get({'_id': ObjectId(post_id)}))
    except Post.DoesNotExist:
        response = web.HTTPNotFound(
            text=aiohttp_jinja2.render_string('404.html', request, {}))
        response.content_type = 'text/html'
        return response

    return aiohttp_jinja2.render_template(
        'post.html', request, {'post': post})


@aiohttp_jinja2.template('post.html')
async def new_comment(request):
    session = await aiohttp_session.get_session(request)
    form = await request.post()
    post_id = ObjectId(form['post_id'])
    try:
        post = await dereference(
            await Post.objects.get({'_id': post_id}))
    except Post.DoesNotExist:
        session['flash'] = 'No post with id: %s' % post_id
        raise web.HTTPFound(request.app.router['index'].url_for())

    comment = Comment(
        author=form['author'],
        date=datetime.datetime.now(),
        body=form['content'])
    post.comments.append(comment)
    try:
        await post.save()
    except ValidationError as e:
        post.comments.pop()
        comment_errors = e.message['comments'][-1]

        return {
            'post': post,
            'errors': comment_errors
        }

    return {
        'post': post,
        'flashed_messages': ['Comment saved successfully.']
    }


def main():
    loop = asyncio.get_event_loop()
    app = loop.run_until_complete(init_app(loop))
    web.run_app(app, host='127.0.0.1', port=5000)


if __name__ == '__main__':
    main()
